fuzz.go

```diff
@@ -1,90 +1,139 @@
 package p2p
 
 import (
+	"fmt"
 	"math/rand"
 	"net"
 	"sync"
 	"time"
-
-	cfg "github.com/tendermint/go-config"
 )
 
-//--------------------------------------------------------
-// delay reads/writes
-// randomly drop reads/writes
-// randomly drop connections
-
 const (
-	FuzzModeDrop  = "drop"
-	FuzzModeDelay = "delay"
+	// FuzzModeDrop is a mode in which we randomly drop reads/writes, connections or sleep
+	FuzzModeDrop = iota
+	// FuzzModeDelay is a mode in which we randomly sleep
+	FuzzModeDelay
 )
 
-func FuzzConn(config cfg.Config, conn net.Conn) net.Conn {
+type FuzzConnConfig struct {
+	Mode         int
+	MaxDelay     time.Duration
+	ProbDropRW   float64
+	ProbDropConn float64
+	ProbSleep    float64
+}
+
+func defaultFuzzConnConfig() *FuzzConnConfig {
+	return &FuzzConnConfig{
+		Mode:         FuzzModeDrop,
+		MaxDelay:     3 * time.Second,
+		ProbDropRW:   0.2,
+		ProbDropConn: 0.00,
+		ProbSleep:    0.00,
+	}
+}
+
+func FuzzConn(conn net.Conn) net.Conn {
+	return FuzzConnFromConfig(conn, defaultFuzzConnConfig())
+}
+
+func FuzzConnFromConfig(conn net.Conn, config *FuzzConnConfig) net.Conn {
+	return &FuzzedConnection{
+		conn:         conn,
+		start:        make(<-chan time.Time),
+		active:       true,
+		mode:         config.Mode,
+		maxDelay:     config.MaxDelay,
+		probDropRW:   config.ProbDropRW,
+		probDropConn: config.ProbDropConn,
+		probSleep:    config.ProbSleep,
+	}
+}
+
+func FuzzConnAfter(conn net.Conn, d time.Duration) net.Conn {
+	return FuzzConnAfterFromConfig(conn, d, defaultFuzzConnConfig())
+}
+
+func FuzzConnAfterFromConfig(conn net.Conn, d time.Duration, config *FuzzConnConfig) net.Conn {
 	return &FuzzedConnection{
-		conn:   conn,
-		start:  time.After(time.Second * 10), // so we have time to do peer handshakes and get set up
-		params: config,
+		conn:         conn,
+		start:        time.After(d),
+		active:       false,
+		mode:         config.Mode,
+		maxDelay:     config.MaxDelay,
+		probDropRW:   config.ProbDropRW,
+		probDropConn: config.ProbDropConn,
+		probSleep:    config.ProbSleep,
 	}
 }
 
+// FuzzedConnection wraps any net.Conn and depending on the mode either delays
+// reads/writes or randomly drops reads/writes/connections.
 type FuzzedConnection struct {
 	conn net.Conn
 
-	mtx   sync.Mutex
-	fuzz  bool // we don't start fuzzing right away
-	start <-chan time.Time
+	mtx    sync.Mutex
+	start  <-chan time.Time
+	active bool
 
-	// fuzz params
-	params cfg.Config
+	mode         int
```

Comment:
don't know. I was looking into https://godoc.org/golang.org/x/net/websocket#Conn and it is not clear whenever the config is a field of `Conn`. But yeah, let's do this.